flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlCreateMaterializedTable.java

```diff
@@ -150,20 +170,33 @@ public SqlNode getAsQuery() {
         return asQuery;
     }
 
+    /** Returns the column constraints plus the table constraints. */
+    public List<SqlTableConstraint> getFullConstraints() {
+        return SqlConstraintValidator.getFullConstraints(tableConstraints, columnList);
+    }
+
+    @Override
+    public void validate() throws SqlValidateException {
+        if (!isSchemaWithColumnsIdentifiersOnly()) {
+            SqlConstraintValidator.validateAndChangeColumnNullability(tableConstraints, columnList);
+        }
+    }
+
+    public boolean isSchemaWithColumnsIdentifiersOnly() {
+        // CREATE MATERIALIZED TABLE supports passing only column identifiers in the column list. If
+        // the first column in the list is an identifier, then we assume the rest of the
+        // columns are identifiers as well.
+        return !getColumnList().isEmpty() && getColumnList().get(0) instanceof SqlIdentifier;
+    }
+
     @Override
     public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
         writer.keyword("CREATE MATERIALIZED TABLE");
         tableName.unparse(writer, leftPrec, rightPrec);
 
-        if (tableConstraint != null) {
-            writer.newlineAndIndent();
+        if (!columnList.isEmpty() || !tableConstraints.isEmpty() || watermark != null) {
```

Comment:
I am curious - can we have a watermark when there are no columns? I am not sure what a no column materialized table means - is this allowed?